src/ratelimiter_map.rs

```diff
@@ -0,0 +1,135 @@
+        Duration::from_secs(var("CLIENT_REAP_INTERVAL").map_or(600, |timeout| {
+            if let Ok(timeout_secs) = timeout.parse() {
+                timeout_secs
+            } else {
+                warn!("Unable to parse CLIENT_REAP_INTERVAL, proceeding with defaults");
+                600
+            }
+        }));
+
+    let client_decay_timeout =
+        Duration::from_secs(var("CLIENT_DECAY_TIEOUT").map_or(3600, |timeout| {
+            if let Ok(timeout_secs) = timeout.parse() {
+                timeout_secs
+            } else {
+                warn!("Unable to parse CLIENT_DECAY_TIMEOUT, proceeding with defaults");
+                3600
+            }
+        }));
+
+    let mut interval = interval(client_reap_interval);
+
+    loop {
+        interval.tick().await;
+        let right_now = Instant::now();
+
+        map.retain(|_, (_, last_used)| *last_used + client_decay_timeout > right_now);
+
+        debug!("Done reaping timed out HTTP ratelimiters");
+    }
+}
+
+impl RatelimiterMap {
+    pub fn new(mut default_token: String) -> Self {
+        let is_bot = default_token.starts_with("Bot ");
+        let is_bearer = default_token.starts_with("Bearer ");
+
+        // Make sure it is either a bot or bearer token, and assume it's a bot
+        // token if no prefix is given
+        if !is_bot && !is_bearer {
+            default_token.insert_str(0, "Bot ");
+        }
+
+        let max_size = var("CLIENT_CACHE_MAX_SIZE").map_or(None, |size| {
+            if let Ok(size) = size.parse() {
+                Some(size)
+            } else {
+                warn!("Unable to parse CLIENT_CACHE_MAX_SIZE, proceeding with defaults");
+                None
+            }
+        });
+
+        let inner = Arc::new(DashMap::new());
+        let default = Ratelimiter::new();
+
+        tokio::spawn(reap_old_ratelimiters(inner.clone()));
+
+        Self {
+            default,
+            default_token,
+            max_size,
+            inner,
+        }
+    }
+
+    fn lru(&self) -> Option<RefMulti<String, (Ratelimiter, Instant)>> {
+        self.inner.iter().next().map(|first_entry| {
+            self.inner.iter().fold(
+                first_entry,
+                |old, next| {
+                    if old.1 > next.1 {
+                        next
+                    } else {
+                        old
+                    }
+                },
+            )
+        })
+    }
+
+    pub fn get_or_insert(&self, token: Option<&str>) -> (Ratelimiter, String) {
+        if let Some(token) = token {
+            if token == self.default_token {
+                (self.default.clone(), self.default_token.clone())
+            } else {
+                let access_time = Instant::now();
+
+                if let Some(mut entry) = self.inner.get_mut(token) {
+                    entry.1 = access_time;
+                    (entry.0.clone(), token.to_string())
+                } else {
+                    if self
```

Comment:
Using `Option::filter` is weird here, using something like `map` may be preferable.